### ISSUE TITLE
[WebProfiler] don't add inline javascript

### ISSUE THREAD

**Issue by thkoch2001** [symfony/symfony#15397]
Inline JavaScript (and CSS) is a security and performance issue. Content-Security-Policy exists to tell browsers not to execute inline JavaScript.

The Web Profiler Toolbar however uses inline JavaScript. Why? It would also be possible to add a script tag to load the missing JavaScript.

Informations can be passed from the server to the JavaScript code easily without inline JavaScript:
```html
<script class="embedded-json-data" type="application/json" data-name="myActiveProfile">
   {"id": 123, "name": "ido", "language": "en"}
</script>
```
and in your JavaScrip (angular here)t:
```js
var selector = 'script.embedded-json-data[data-name="' + name + '"]',
    node = document.querySelector(selector),
    data = angular.fromJson(node.innerHTML);
```

* [Datenübergabe innerhalb Webapplikationen](http://www.comsolit.com/blog/2014/04/datenuebergabe-innerhalb-webapplikationen/) (German)
* https://mathiasbynens.be/notes/json-dom-csp

**Comment by jakzal** [comment]
Does it really matter in case of the web toolbar which is meant to be used in development only?

**Comment by stof** [comment]
@jakzal this forces you to disable CSP headers during development, meaning that any CSP-incompatible code would only be detected in prod.

**Comment by stof** [comment]
I have some idea about how this could be done.
This would involve a big change in the WebProfilerBundle, and require bundles to update their code to register their JS and custom styles in a new way (otherwise, they would continue to work as long as you don't have CSP in place, but would not allow CSP). But thanks to the new system we put in places in the WebProfilerBundle to allow bundles to support different versions of the profiler in parallel while knowing which one is used, it is now possible to introduce such refactoring in a BC way IMO (I need to confirm it though to be sure).
I have no time to work on this before the feature freeze of Symfony 2.8 and 3.0 (which is at the end of the month), but I'm putting this on my todo-list for Symfony 3.1.

**Comment by vooj-tae** [comment]
For now, I am using this workaround (Symfony 2.7) - in **Kernel Response event listener** with lower priority than _"WebDebugToolbarListener"_, I am adding a CSP "nonce" exception like this:

    if ($this->kernel->getEnvironment() == "dev") {
        $crawler = new Crawler($response->getContent());
        $scripts = $crawler->filter('script');
        foreach($scripts as $script) {
            if ($script->textContent) {
                if (strpos($script->textContent, "Sfjs") !== false) {
                    $nonce = hash('sha256', random_bytes(32));
                    $script->setAttribute("nonce", $nonce);
                    $security_policy["script-src"][] = "'nonce-".$nonce."'";
                }
            }
        }
        $response->setContent($crawler->html());
    }

`$security_policy` is what I then parse into **Content-Security-Policy** HTTP header. There is probably a better/safer way to identify toolbar scripts than searching for "Sfjs", but this is executed only for development, so (as @stof pointed out) you don't have to disable CSP during development and detect incompatibilities in dev.

**Comment by stof** [comment]
@vooj-tae see https://github.com/symfony/symfony/pull/18568 for the PR implementing support for CSP in the toolbar (this still requires disabling CSP for the profiler pages themselves, which is OK IMO and makes things much easier than what I thought about in September 2015 (no need to change external bundles hooking into the profiler to add new panels)
